[GXA-3518] feat: add Stratified splitters

ludovico-lanni opened the pull request.

david26694 commented on tests/test_splitter.py:
```diff
@@ -106,3 +149,103 @@ def test_switchback_balanced_splitter_abc():
+
+
+def test_stratified(df_strata):
+    splitter = StratifiedClusteredSplitter(
```
There is a lot of code repetition in here, we could extract a function cllaed assert_balanced_strata(df) and just call this function in the tests

ludovico-lanni replied:
done! changed the assertion pipeline to test clusters-strata unique combinations, as this is what the splitter does.

david26694 commented on cluster_experiments/random_splitter.py:
```diff
@@ -198,3 +193,59 @@ def assign_treatment_df(
+        # random shuffling
+        random_sorted_treatments = list(np.random.permutation(self.treatments))
+
+        df_unique_shuffled = df_unique.sample(frac=1)
```
Couple things: you could do this in line 215, in a more pipeline manner, and also do a reset_index(drop=True) after this to have less noise with indices

ludovico-lanni replied:
done!

ludovico-lanni commented on tests/test_splitter.py:
```diff
@@ -106,3 +149,103 @@ def test_switchback_balanced_splitter_abc():
+            ) == 0.5
+
+
+def test_stratified_shuffled_input(df_strata):
```
done! added for other tests as well

ludovico-lanni commented on tests/test_splitter.py:
```diff
@@ -32,6 +33,48 @@ def df_cluster(clusters):
     )
 
 
+@pytest.fixture
```
done!

ludovico-lanni commented on cluster_experiments/random_splitter.py:
```diff
@@ -198,3 +193,59 @@ def assign_treatment_df(
         return cls(treatments=config.treatments, treatment_col=config.treatment_col)
+
+
+class StratifiedClusteredSplitter(RandomSplitter):
```
done!

ludovico-lanni commented:
- refactored code as suggested
- update readme and init files with new class
- changed the logic of the tests as we were testing for balance in the number of rows across strata, instead of number of clusters across strata

david26694 commented on tests/test_splitter.py:
```diff
@@ -106,3 +168,138 @@ def test_switchback_balanced_splitter_abc():
+        strata_cols + cluster_cols + ["treatment"]
+    ].drop_duplicates()
+
+    ##
```
remove unneeded #

ludovico-lanni replied:
done!

david26694 commented on tests/test_splitter.py:
```diff
@@ -106,3 +168,138 @@ def test_switchback_balanced_splitter_abc():
+    ].drop_duplicates()
+
+    ##
+    if len(cluster_cols) > 1:
```
can we extract this into a function? the whole if: agg else

ludovico-lanni replied:
done! `combine_columns(df, cols_list)`

david26694 commented on tests/test_splitter.py:
```diff
@@ -106,3 +168,138 @@ def test_switchback_balanced_splitter_abc():
+
+
+def test_stratified(df_strata):
+
```
style thing: no need for a line in here

ludovico-lanni replied:
done!

david26694 commented on tests/test_splitter.py:
```diff
@@ -106,3 +168,138 @@ def test_switchback_balanced_splitter_abc():
+def test_stratified(df_strata):
+
+    # base test for balance of treatments across strata
+
```
style thing: no need for a line in here

ludovico-lanni replied:
done!

david26694 reviewed (approved):
could you change these two minor things and merge afterwards?

ludovico-lanni commented:
- fixed minor style inaccuracies
- extracted function to combine columns
